fgpyo/samtools.py

```diff
@@ -0,0 +1,375 @@
+        sort_unmapped_reads: If true, sort unmapped reads by their sequence minimizer, reverse
+            complementing where appropriate. This has the effect of collating some similar data
+            together, improving the compressibility of the unmapped sequence. The minimiser kmer
+            size is adjusted using the ``kmer_size`` option. Note data compressed in this manner
+            may need to be name collated prior to conversion back to fastq.
+
+            Mapped sequences are sorted by chromosome and position.
+        kmer_size: the kmer-size to be used if sorting unmapped reads.
+        compression_level: The compression level to be used in the final output file.
+        memory_per_thread: Approximately the maximum required memory per thread, specified either
+            in bytes or with a K, M, or G suffix.
+
+            To prevent sort from creating a huge number of temporary files, it enforces a minimum
+            value of 1M for this setting.
+        sort_order: The sort order to use when sorting the file.
+        sort_tag: The tag to use to use to sort the SAM/BAM/CRAM records. Will be sorted by
+            this tag first, followed by position (or name, depending on ``sort_order``
+            provided).
+        output_format: the output file format to write the results as.
+            By default, will try to select a format based on the ``output`` filename extension;
+            if no format can be deduced, bam is selected.
+        tempfile_prefix: The prefix to use for temporary files. Resulting files will be in
+            format PREFIX.nnnn.bam, or if the specified PREFIX is an existing directory, to
+            PREFIX/samtools.mmm.mmm.tmp.nnnn.bam, where mmm is unique to this invocation of the
+            sort command.
+
+            By default, any temporary files are written alongside the output file, as
+            out.bam.tmp.nnnn.bam, or if output is to standard output, in the current directory
+            as samtools.mmm.mmm.tmp.nnnn.bam.
+        threads: The number of threads to use when sorting. By default, operation is
+            single-threaded.
+        no_pg: If true, will not add a @PG line to the header of the output file.
+    """
+
+    output_string = (
+        f"{output}##idx##{output}{output_format.index_extension}"
+        if index_output and output_format.index_extension is not None
+        else str(output)
+    )
+
+    args = ["-m", memory_per_thread, "-O", output_format._name_, "-@", str(threads)]
+
+    if sort_unmapped_reads:
+        args.extend(["-M", "-K", str(kmer_size)])
+
+    if compression_level is not None:
+        args.extend(["-I", str(compression_level)])
+
+    if sort_order == SamOrder.QueryName:
+        args.append("-n")
+    elif sort_order == SamOrder.TemplateCoordinate:
+        args.append("--template-coordinate")
+    else:
+        assert (
+            sort_order == SamOrder.Coordinate
```

Comment:
I think the check should be for "not unknown" and "not unsorted" in case other orders are added.